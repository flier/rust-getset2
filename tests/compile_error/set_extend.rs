use getset2::{Getter, Setter};

#[derive(Getter, Setter)]
pub struct Struct {
    #[get(pub, copy)]
    #[set(pub, extend)]
    // #[set(extend)] is not allowed for a field that is not implements Extend trait.
    field: usize,
}

#[derive(Getter, Setter)]
#[get(pub, copy)]
#[set(pub, extend)] // #[set(extend)] can not be applied to the structure
pub struct Struct2 {
    field: usize,
}

fn main() {}
