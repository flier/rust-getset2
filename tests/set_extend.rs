use std::collections::HashMap;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};

use getset2::{Getter, Setter};

#[derive(Default, Getter, Setter)]
pub struct Foo<'a> {
    #[get(str)]
    #[set(extend)]
    string_field: String,

    #[get(slice)]
    #[set(extend)]
    vec_field: Vec<usize>,

    #[set(extend)]
    map_field: HashMap<usize, usize>,

    #[set(extend(&'a Path))]
    path_field: PathBuf,

    #[set(extend(P: AsRef<Path>))]
    p_field: PathBuf,

    #[get(skip)]
    #[set(skip)]
    phantom: PhantomData<&'a u8>,
}

#[test]
fn set_extend() {
    let mut foo = Foo::default();

    assert_eq!(
        foo.extend_string_field("foo".chars())
            .extend_string_field("bar".chars())
            .append_string_field('!')
            .string_field(),
        "foobar!"
    );

    assert_eq!(
        foo.extend_vec_field([1, 2, 3])
            .extend_vec_field([4, 5, 6])
            .append_vec_field(7)
            .vec_field(),
        [1, 2, 3, 4, 5, 6, 7]
    );

    assert_eq!(
        foo.extend_map_field(vec![(1, 2), (3, 4)])
            .append_map_field((5, 6))
            .map_field()
            .get(&5)
            .unwrap(),
        &6
    );

    assert_eq!(
        foo.extend_path_field([Path::new("/"), Path::new("foo")])
            .append_path_field(Path::new("bar"))
            .path_field(),
        Path::new("/foo/bar")
    );

    assert_eq!(
        foo.extend_p_field(["/", "foo"])
            .append_p_field("bar")
            .p_field(),
        Path::new("/foo/bar")
    )
}
