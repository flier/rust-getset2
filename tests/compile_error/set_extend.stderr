error: #[set(extend)] supports only some of the well-known types,
       #[set(extend(Item))] should be used for a type which implements the `Extend<Item>` trait
 --> tests/compile_error/set_extend.rs:8:12
  |
8 |     field: usize,
  |            ^^^^^

error: invalid #[set(..)] attribute, cannot find parameter `extend` in this scope
  --> tests/compile_error/set_extend.rs:13:1
   |
13 | #[set(pub, extend)] // #[set(extend)] can not be applied to the structure
   | ^
